libsrc/common/getcpu.s

```diff
@@ -26,52 +27,118 @@
 .p816                           ; Enable 65816 instructions
 
 _getcpu:
+
         lda     #0
         inc     a               ; .byte $1A ; nop on nmos, inc on every cmos
         cmp     #1
-        bcc     @L8
+        bcc     @IsNMOS
 
 ; This is at least a 65C02, check for a 65CE02/4510
 
         .byte   $42,$EA         ; neg on 65CE02/4510, nop #$EA on 65C02, wdm $EA on 65816
         cmp     #1
-        beq     @L6
+        beq     @HasINCA
 
 ; This is at least a 65CE02, check for 4510
 
         lda     #5              ; CPU_65CE02 constant
+	ldx	#0		; To make sure MAP doesn't do anything, the upper nybl of X and Z must be clear
         .byte   $5C             ; map on 4510, aug on 65CE02 (acts like 4 byte nop)
         lda     #3              ; CPU_4510 constant
         nop
-        bne     @L9
+	cmp	#5
+        beq     @LoadXAndReturn
+
+; It is either a 4510 (C65) or a 45GS02 (MEGA65)
+
+        ; 45GS02 supports 32-bit ZP indirect, so use that to check CPU type
+        ; without requiring a functioning MEGA65 hypervisor.
+        ; We setup a read of $200FF, then store a different value in $00FF
+        ; and then re-read $200FF to see if it is unchanged.
+
+        ; Save the 32-bit ZP pointer and data byte
+        ldx     #$04
+@L10:   lda     $fb,x
+        sta     @GetCPUTemp,x
+        dex
+        bpl     @L10
+
+        ; Setup 32-bit pointer
+        lda     #$ff
+        sta     $fb
+        lda     #$00
+        sta     $fc
+        sta     $fe
+        lda     #$02
+        sta     $fd
+
+        ; Prefixing LDA ($nn),Z with a NOP uses 32-bit ZP pointer on 45GS02,
+        ; but normal 16-bit ZP pointer on 4510
+        ; (We assume Z=$00, which will be the normal case)
+        nop                     ; prefix to tell next instruction to be 32-bit ZP
+        .byte   $b2,$fb         ; LDA ($nn),Z
+        eor     #$ff            ; change the value
+        sta     $ff             ; store in $FF
+        ; now try again to load it: If the same, then 45GS02, as $200FF is unchanged
+        nop                     ; prefix to tell next instruction to be 32-bit ZP
+        .byte   $b2,$fb         ; LDA ($nn),Z
+        cmp     $ff             ; does the loaded value match what is in $FF?
+        beq     @Is4510         ; matches, so must be a 4510 = C65
+        bne     @Is45GS02       ; $200FF and $FF have different values, so must be a MEGA65 45GS02
+@Is4510:
+        jsr     @RestoreGetCPUTemp
+        lda     #$03            ; CPU_4510 constant
+        ldx     #0              ; Load high byte of word
```

Comment:
Please decide on #0 vs. #$0 vs. #$00 coding style and be consistent about it.

Reply:
Please decide on starting comments with upper case letter or not and be consistent about it.